### ISSUE TITLE
Project planing

### ISSUE THREAD

**Issue by reygoch** [FER-PUH/dormouse#1]
We should start by defining what features should be available in the new system. Let's not go overboard with suggestions since people loose their motivation very quickly.

Idea is to define the bare minimum set of features first and from that we should define our data entities / model. After that is done we can move on to actual coding and everything else will be left to the future PUH students to improve upon :D

Here are some features I've scribbled down during the last meeting:

1. Users
2. Roles for the current year / iteration of PUH (user can be student one year and TA the next)
3. Homework upload
4. Grading
5. News / updates
6. Questions
7. Peer review
8. Automatic homework testing on upload
9. Email notifications

From those I think we can already exclude 8 and 6.

**Comment by reygoch** [comment]
I've sketched a little ER diagram for our system. I'll try to write specifications tomorrow. You should check out the diagram and comment if there is something illogical or if you have something to add (or remove).

![dormouse-er](https://user-images.githubusercontent.com/8524934/36951444-3b632358-2004-11e8-8619-ec04930f2b5f.png)

**Comment by reygoch** [comment]
Ok, so here is the deal.

The basis of everything is the **Year**. Administrator should be able to define a new academic year in the system. We also have the **Users** which can register account independently of the year but can't really do much unless they have been assigned a **Role** for a certain **Year**.

This is necessary because single user can be a student one year and TA next year and if he becomes TA he shouldn't be able to have TA view (see details, student scores etc.) of the previous years but only for the year when he is a TA (and he should also have access to his personal scores and files from the year when he was a student).

Basically we will have two roles: **Student** and **TA** and those roles should be assigned to users on a yearly basis.

Since the main purpose of the system is **Task** submission and grading **TAs** can create new **Tasks** that can have following elements:

+ Title
+ Description
+ Attachments
+ Deadlines
+ Elements

**Attachments** can be PDF containing more detailed task description, or a ZIP containing some resources needed to complete the task.

**Deadlines** will define the deadlines for completing the task, and each deadline will also have a factor that will multiply final score of the student. This is so that we can give for example only 60% of maximum score if the student has gotten full points but submitted his solution on the second deadline.

**Elements** will define expected file uploads from the students (e.g. Homework.hs and Exercises.hs) and each element will have defined number of points an can be graded and commented on by the TA.

Students can upload multiple solutions for each task, but only the last one will be presented to the TA for grading. We will have **Submission** table which is basically just a linking table between **Student** and **Files** for a certain task.

For every submission we can have multiple reviews which are represented by **Review** table and every review is composed of multiple elements / file uploads that will be graded and commented on individually.

---

I hope that this makes sense :D

**Comment by jasminzukic** [comment]
Makes sense.

I would also like if TAs could have access to previous tasks from earlier years. The ones that were published but also tasks that were recommended but didn't find themselves in the final PDF.

So TAs will have additional access to that element of previous years.

Also for those tasks to have tags in them which contain info on how far in the lectures students have to be to be able to implement them (e.g. A task on ADTs won't be put in the PDF if the students covered recursions last lecture). I hope I've been clear about this.

**Comment by reygoch** [comment]
@jasminzukic tasks from previous year should be freely available for everyone to see (download PDF and extra files) it's just grades and student submissions that aren't visible to you if you weren't TA that year.

I'm not sure about adding tags as well, since we are trying to achieve minimal working example. I'm thinking that this Admin table should be nixed as well and instead users will have column "admin" with true or false.

We might do tags just as a simple text field within the Task table?

**Comment by jasminzukic** [comment]
Yeah, that can be additional work. Not necessary for the minimum. Just to have a database of all tasks (from PDF-s and recommended ones that didn't get into the PDF)

**Comment by reygoch** [comment]
Ok, I think this should be enough for us to get started with writing an actual database schema. I'll prepare some boilerplate so that we can easily (relative term :) create necessary database users / tables etc and be in sync.
